File: CoreServer/datamodel/server/datamodelmanager.cpp
#include "datamodelmanager.h"

#include <QDebug>

#include "datamodel/server/httpdatamodelloader.h"
#include "datamodel/server/testdatamodelloader.h"
#include "datamodel/server/filedatamodelloader.h"
#include "value/server/servervaluemanager.h"
#include "actor/actormanager.h"
#include "datamodel/server/emptydatamodel.h"

#include "macros.h"

QLatin1String DatamodelManager::MANAGER_ID = QLatin1String("DatamodelManager");

DatamodelManager::DatamodelManager(QObject *parent) : ManagerBase(parent)
{
    m_datamodel = new EmptyDatamodel();
}

LogCat::LOGCAT DatamodelManager::logCat() {
    return LogCat::DATAMODEL;
}

void DatamodelManager::init(LocalConfig* config) {
    iDebug() << Q_FUNC_INFO;

    REQUIRE_MANAGER(ValueManagerBase);
    REQUIRE_MANAGER(ActorManager);

    QString datamodelLoaderName = config->getString("datamodel.loader", TestDatamodelLoader::LOADER_TYPE_NAME);

    if (datamodelLoaderName == HttpDataModelLoader::LOADER_TYPE_NAME) {
        m_datamodelLoader = new HttpDataModelLoader(QUrl(config->getString("datamodel.url", "http://localhost/datamodel")));
    } else if (datamodelLoaderName == FileDataModelLoader::LOADER_TYPE_NAME) {
        m_datamodelLoader = new FileDataModelLoader(config->getString("datamodel.filePath", "datamodel.json"));
    } else if (datamodelLoaderName == TestDatamodelLoader::LOADER_TYPE_NAME) {
        m_datamodelLoader = new TestDatamodelLoader();
    } else {
        iWarning() << "Unknown datamodel loader" << datamodelLoaderName;
    }

    Q_ASSERT(m_datamodelLoader != nullptr);

    connect(m_datamodelLoader, &DatamodelLoaderBase::saved, this, &DatamodelManager::onDatamodelSaved);
    connect(m_datamodelLoader, &DatamodelLoaderBase::error, this, &DatamodelManager::onDatamodelError);

    delete m_datamodel;
    m_datamodel = m_datamodelLoader->load();
    Q_ASSERT(m_datamodel != nullptr);

    m_isLoaded = true;

    iDebug() << "Datamodel loaded";

    Q_EMIT(datamodelChanged());

    registerValues();

    registerActors();
}

void DatamodelManager::registerValues() {
    // register values
    ValueManagerBase* valueManager = getManager<ValueManagerBase>(ValueManagerBase::MANAGER_ID);

    QMapIterator<QString, ValueBase*> it(m_datamodel->values());
    while (it.hasNext()) {
        it.next();
        iDebug() << "Register datamodel value" << it.key();
        valueManager->registerValue(it.value());
    }
}

void DatamodelManager::registerActors() {
    // register actors
    ActorManager* actorManager = getManager<ActorManager>(ActorManager::MANAGER_ID);
    ValueManagerBase* valueManager = getManager<ValueManagerBase>(ValueManagerBase::MANAGER_ID);

    QMapIterator<QString, ActorBase*> it(m_datamodel->actors());
    while(it.hasNext()) {
        it.next();
        iDebug() << "Register datamodel actor" << it.key();
        actorManager->registerActor(it.value());
        valueManager->registerValue(it.value());
    }
}

QString DatamodelManager::id() {
    return MANAGER_ID;
}

MessageBase::MESSAGE_TYPE DatamodelManager::getMessageType() {
    return MessageBase::MESSAGE_TYPE_UNKNOWN;
}

void DatamodelManager::handleReceivedMessage(MessageBase* msg) {
}

DatamodelBase* DatamodelManager::datamodel() {
    return m_datamodel;
}

bool DatamodelManager::isLoaded() {
    return m_isLoaded;
}

void DatamodelManager::onDatamodelSaved() {
    iDebug() << "Datamodel saved";
}

void DatamodelManager::onDatamodelError(QString desc) {
    iWarning() << "Error while saving datamodel" << desc;
}

void DatamodelManager::save() {
    iDebug() << Q_FUNC_INFO;

    m_datamodelLoader->save(m_datamodel);
}

File: AndroidApp/OshCore/src/main/java/com/osh/datamodel/DatamodelManager.java
package com.osh.datamodel;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.osh.actor.IActorManager;
import com.osh.communication.MessageBase;
import com.osh.communication.MessageBase.MESSAGE_TYPE;
import com.osh.config.ApplicationConfig;
import com.osh.manager.IManagerRegistration;
import com.osh.manager.ManagerBase;
import com.osh.service.IDatamodelService;
import com.osh.value.IValueManager;

public class DatamodelManager extends ManagerBase implements IDatamodelManager {

	private static final Logger log = LoggerFactory.getLogger(DatamodelManager.class);

    private DatamodelLoaderBase datamodelLoader;
    private DatamodelBase datamodel;

	private ApplicationConfig appConfig;
	
	private IValueManager valueManager;
	
	private IActorManager actorManager;
	
	private IDatamodelService datamodelService;

    public DatamodelManager(IManagerRegistration managerRegistration) {
		super("DatamodelManager", managerRegistration);
	}

	@Override
	public MESSAGE_TYPE getMessageType() {
		return MESSAGE_TYPE.MESSAGE_TYPE_UNKNOWN;
	}

	@Override
	public void handleReceivedMessage(MessageBase msg) {
	}

	@Override
	public DatamodelBase getDatamodel() {
		return datamodel;
	}

	public void registerValues() {
		// register values
		for (String key : datamodel.getValues().keySet()) {
			log.debug("Register value {}", key);
			valueManager.registerValue(datamodel.getValues().get(key));
		}
	}

	public void registerActors() {
		// register actors
		for (String key : datamodel.getActors().keySet()) {
			log.debug("Register actor {}", key);
			actorManager.registerActor(datamodel.getActors().get(key));
			valueManager.registerValue(datamodel.getValues().get(key));
		}
	}

	@Override
	public void save() {
		datamodelLoader.save(datamodel);
	}

	@Override
	public void initComplete() {
		log.info("Init Datamodel");
		
	    String datamodelLoaderName = appConfig.getDatamodel().getLoader();

	    if (datamodelLoaderName.equals(HttpDatamodelLoader.LOADER_TYPE_NAME)) {
	        datamodelLoader = new HttpDatamodelLoader(appConfig.getDatamodel().getUrl());
	    //} else if (datamodelLoaderName.equals(FileDataModelLoader.LOADER_TYPE_NAME)) {
	        //datamodelLoader = new FileDataModelLoader(appConfig.getDatamodel().getFilePath());
	    } else if (datamodelLoaderName.equals(TestDatamodelLoader.LOADER_TYPE_NAME)) {
	        datamodelLoader = new TestDatamodelLoader();
	    } else if (datamodelLoaderName.equals(DBDatamodelLoader.LOADER_TYPE_NAME)) {
	    	datamodelLoader = new DBDatamodelLoader(datamodelService, appConfig.getDatamodel().getModelName());
	    } else {
	        log.warn("Unknown datamodel loader {}", datamodelLoaderName);
	    }
	    

	    /*
	    DBDatamodel model = new DBDatamodel("default");
	    ValueGroup vg = model.addValueGroup("testGroup");
	    model.addDigitalActor(vg, "act1", ValueType.VT_BRIGHTNESS, false, VALUE_TIMEOUT.VT_LONG);
	    model.addBooleanValue(vg, "test123", ValueType.VT_MOTION, VALUE_TIMEOUT.VT_NONE);
	    FileDataModelLoader testfd = new FileDataModelLoader("test.json");
	    testfd.save(model);
	    datamodelLoader.save(model);
	    */
	    
	    datamodel = datamodelLoader.load();

	    log.info("Datamodel loaded");

	    registerValues();

	    //registerActors();
	}

}
